src/celeritas/optical/interactor/RayleighInteractor.hh

```diff
@@ -0,0 +1,111 @@
+#include "celeritas/random/distribution/IsotropicDistribution.hh"
+
+namespace celeritas
+{
+namespace optical
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Sample optical Rayleigh scattering.
+ *
+ * Optical Rayleigh scattering is the elastic scattering of optical photons
+ * in a material. The scattered polarization is guarenteed to be in the
+ * same plane as the original polarization and new direction if the latter
+ * vectors are not parallel. Otherwise it will just be perpendicular to
+ * the new direction.
+ */
+class RayleighInteractor
+{
+  public:
+    //! Construct interactor from an optical track
+    inline CELER_FUNCTION
+    RayleighInteractor(TrackView const& particle, Real3 const& direction);
+
+    //! Sample an interaction with the given RNG
+    template<class Engine>
+    inline CELER_FUNCTION Interaction operator()(Engine& rng) const;
+
+  private:
+    Real3 const& inc_dir_;  //!< Direction of incident photon
+    Real3 const& inc_pol_;  //!< Polarization of incident photon
+};
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Construct the interactor for the given optical track.
+ */
+CELER_FUNCTION RayleighInteractor::RayleighInteractor(TrackView const& particle,
+                                                      Real3 const& direction)
+    : inc_dir_(direction), inc_pol_(particle.polarization())
+{
+    CELER_EXPECT(is_soft_unit_vector(inc_dir_));
+    CELER_EXPECT(is_soft_unit_vector(inc_pol_));
+    CELER_EXPECT(soft_zero(dot_product(inc_dir_, inc_pol_)));
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Sample the scattered direction and polarization for a single optical
+ * Rayliegh interaction.
+ */
+template<class Engine>
+CELER_FUNCTION Interaction RayleighInteractor::operator()(Engine& rng) const
+{
+    Interaction result;
+    Real3& new_dir = result.direction;
+    Real3& new_pol = result.polarization;
+
+    do
+    {
+        new_dir = IsotropicDistribution{}(rng);
+
+        auto projected_pol = dot_product(new_dir, inc_pol_);
+
+        if (projected_pol == 0)
+        {
+            // If new direction is parallel to incident polarization, then
+            // randomly sample azimuthal direction
+            new_pol = ExitingDirectionSampler{0, new_dir}(rng);
+        }
+        else
+        {
+            // Project polarization onto plane perpendicular to new direction
+            new_pol = make_unit_vector(inc_pol_ - projected_pol * new_dir);
+            new_pol *= BernoulliDistribution{real_type{0.5}}(rng) ? 1 : -1;
+        }
+
+        // Perform rejection sampling for with the probability of the
+        // scattered polarization overlap squared
+    } while (
+        BernoulliDistribution{ipow<2>(dot_product(new_pol, inc_pol_))}(rng));
```

Comment:
Should this be
```suggestion
        !BernoulliDistribution{ipow<2>(dot_product(new_pol, inc_pol_))}(rng));
```
?

Reply:
Sorry if I'm being dense, but isn't `ipow<2>(dot_product(new_pol, inc_pol_))` equal to $\cos^2 \theta$? Which would mean we're currently _rejecting_ with probability $\cos^2 \theta$ (since `BernoulliDistribution{p_true}(rng)` is equivalent to `u < p_true`)?

Reply:
Perhaps this is my fault for insisting on using the non intuitive "Bernoulli distribution". Perhaps a helper function that does "reject less than" would be more readable and less error prone. We've already caught one or two of these...

Reply:
> Sorry if I'm being dense, but isn't `ipow<2>(dot_product(new_pol, inc_pol_))` equal to cos2⁡θ? Which would mean we're currently _rejecting_ with probability cos2⁡θ (since `BernoulliDistribution{p_true}(rng)` is equivalent to `u < p_true`)?

Nope you're absolutely right, I mixed up that passing the Bernoulli sample was a rejection instead of accepting

Reply:
Yeah we’ve run into this rejection/acceptance confusion with the Bernoulli distribution more than a few times now… you've said before we need a rejection sampler, I think you're right :)

Reply:
Yep https://github.com/celeritas-project/celeritas/pull/1138#pullrequestreview-1914662560 . I'll put that together right quick...

Reply:
https://github.com/celeritas-project/celeritas/pull/1354